connector/jobrunner/runner.py

```diff
@@ -0,0 +1,303 @@
+            # TODO: perhaps we don't need to trigger ON DELETE?
+            cr.execute("""
+                DROP TRIGGER IF EXISTS queue_job_notify ON queue_job;
+
+                CREATE OR REPLACE
+                    FUNCTION queue_job_notify() RETURNS trigger AS $$
+                DECLARE
+                    uuid TEXT;
+                BEGIN
+                    IF TG_OP = 'DELETE' THEN
+                        uuid = OLD.uuid;
+                    ELSE
+                        uuid = NEW.uuid;
+                    END IF;
+                    PERFORM pg_notify('connector',
+                                      current_database() || ',' || uuid);
+                    RETURN NULL;
+                END;
+                $$ LANGUAGE plpgsql;
+
+                CREATE TRIGGER queue_job_notify
+                    AFTER INSERT OR UPDATE OR DELETE
+                    ON queue_job
+                    FOR EACH ROW EXECUTE PROCEDURE queue_job_notify();
+            """)
+            cr.execute("LISTEN connector")
+
+    def select_jobs(self, where, args):
+        query = "SELECT %s, uuid, %s, date_created, priority, eta, state " \
+                "FROM queue_job WHERE %s" % \
+                ('channel' if self.has_channel else 'NULL',
+                 'seq' if self.has_seq else 0,
+                 where)
+        with closing(self.conn.cursor()) as cr:
+            cr.execute(query, args)
+            return list(cr.fetchall())
+
+    def set_job_enqueued(self, uuid):
+        with closing(self.conn.cursor()) as cr:
+            cr.execute("UPDATE queue_job SET state=%s, "
+                       "date_enqueued=date_trunc('seconds', now()::timestamp) "
+                       "WHERE uuid=%s",
+                       (ENQUEUED, uuid))
+
+
+class ConnectorRunner:
+
+    def __init__(self, port=8069, channel_config_string='root:4'):
+        self.port = port
+        self.channel_manager = ChannelManager()
+        self.channel_manager.simple_configure(channel_config_string)
+        self.db_by_name = {}
+
+    def get_db_names(self):
+        if openerp.tools.config['db_name']:
+            db_names = openerp.tools.config['db_name'].split(',')
+        else:
+            db_names = openerp.service.db.exp_list()
+        dbfilter = openerp.tools.config['dbfilter']
+        if dbfilter:
+            db_names = [d for d in db_names if re.match(dbfilter, d)]
+        return db_names
+
+    def initialize_databases(self):
+        for db_name, db in self.db_by_name.items():
+            try:
+                self.channel_manager.remove_db(db_name)
+                db.close()
+            except:
+                _logger.warning('error closing database %s',
+                                db_name, exc_info=True)
+        self.db_by_name = {}
+        for db_name in self.get_db_names():
+            db = Database(db_name)
+            if not db.has_connector:
+                _logger.debug('connector is not installed for db %s', db_name)
```

Comment:
When we install a the addon `connector` on a new db, do we need to restart odoo to have it considered by the runner?